device-connectors/src/testflinger_device_connectors/fw_devices/LVFS/LVFS.py

```diff
@@ -0,0 +1,327 @@
+                )
+                if not (rc == 0 or (rc == 1 and "already exists" in stderr)):
+                    raise RuntimeError(
+                        f"[{dev_name}] fail to download firmware file from"
+                        f" LVFS target: {prev_ver['Uri']}\nerror: {stdout}"
+                    )
+                rc, stdout, stderr = self.run_cmd(
+                    f"sudo fwupdmgr install {fw_file} -y "
+                    + "--no-reboot-check --allow-older",
+                    raise_stderr=False,
+                )
+                if rc == 0:
+                    reboot = True
+                else:
+                    logmsg(
+                        logging.ERROR,
+                        f"[{dev_name}] fail to force install (downgrade) "
+                        + f"firmware {fw_file}\n"
+                        + f"error: {stdout}",
+                    )
+            else:
+                logmsg(
+                    logging.INFO,
+                    f"[{dev_name}] not a downgradable component",
+                )
+        return reboot
+
+    def check_results(self) -> bool:
+        """
+        Get upgrade/downgrade result and validate if it succeeds
+
+        :return: `True` if overall upgrade status is success, `False` otherwise
+        """
+        fwupd_result = True
+        for dev in self.fw_info:
+            try:
+                dev_name = dev["Name"]
+            except KeyError:
+                dev_name = dev["DeviceId"]
+            try:
+                expected_ver = dev["targetVersion"]
+            except KeyError:
+                continue
+            rc, stdout, stderr = self.run_cmd(
+                f"sudo fwupdmgr get-results {dev['DeviceId']} --json"
+            )
+            get_results = json.loads(stdout)
+            try:
+                new_fw = get_results["Releases"][0]
+                update_state = get_results["UpdateState"]
+            except KeyError:
+                msg = (
+                    f"[{dev_name}] unable to determine if new firmware "
+                    + "is landed due to missing results"
+                )
+                logmsg(logging.ERROR, msg)
+                fwupd_result = False
+                continue
+
+            if new_fw["Version"] == expected_ver and update_state == 2:
+                msg = (
+                    f"[{dev_name}] firmware flashed {dev['Version']}"
+                    + f" → {expected_ver}"
+                )
+                log_level = logging.INFO
+            else:
+                FwupdUpdateState = [
+                    "FWUPD_UPDATE_STATE_UNKNOWN",
+                    "FWUPD_UPDATE_STATE_PENDING",
+                    "FWUPD_UPDATE_STATE_SUCCESS",
+                    "FWUPD_UPDATE_STATE_FAILED",
+                    "FWUPD_UPDATE_STATE_NEEDS_REBOOT",
+                    "FWUPD_UPDATE_STATE_FAILED_TRANSIENT",
+                    "FWUPD_UPDATE_STATE_LAST",
+                ]
```

Comment:
This might make a really nice place to use an enum. especially when you look at things like line 267 where you have `update_state == 2`. Instead you could do something like this:
(I would define this outside of this method, so that it could be imported by your tests also)
```
from enum import Enum, auto
class FwupdUpdateState(Enum):
        FWUPD_UPDATE_STATE_UNKNOWN = auto(0)
        FWUPD_UPDATE_STATE_PENDING = auto()
        FWUPD_UPDATE_STATE_SUCCESS = auto()
        FWUPD_UPDATE_STATE_FAILED = auto()
        FWUPD_UPDATE_STATE_NEEDS_REBOOT = auto()
        FWUPD_UPDATE_STATE_FAILED_TRANSIENT = auto()
        FWUPD_UPDATE_STATE_LAST = auto()
```
Now, for things like line 267, you can say something more clear like `update_state == FwupdUpdateState.FWUPD_UPDATE_STATE_SUCCESS`. Also, if you need the actual string like on line 285, you can do `... {FwupdUpdateState(update_state).name}` and you'll get the string for the state represented by the index in update_state!

Reply:
Thank you for the suggestion! Using Enum here is really a clever choice. Updated the script and tests accordingly.